### ISSUE TITLE
[Proposal: Creating a high level API for experimentation]

### ISSUE THREAD

**Issue by nipunbatra** [nilmtk/nilmtk#753]
### Goal: Making NILMTK easier for end-users and rapid experimentation

The current method of creating experiments in NILMTK is non-trivial for **end-users** I am proposing a standard configuration that the users can create to support the following use cases:

1. train and test on the same building from the same dataset
2. train and test on different buildings from the same dataset
3. train and test on different buildings across different datasets

Here is a sample configuration that the users can create:

```python
{
	'params': [('power', 'active'), ('power', 'reactive')],
	'sample-rate': 60,
	'appliances': ['fridge', 'air conditioner'],
        'methods': ['FHMM', 'CO'],
        'pre-trained': False,
	'train': {
		'datasets': {
			'REDD': {
				'buildings': [
					'building1': {
						'start-time': '2018-1-1',
						'end-time': '2018-1-2'
					},

					'building2': {
						'start-time': '2018-1-4',
						'end-time': '2018-1-5'
					}
				]
			},
			'IAWE': {
				'buildings': [
					'building1': {
						'start-time': '2018-1-1',
						'end-time': '2018-1-2'
					}
				]
			}
		}
	},
	'test': {
		'datasets': {
			'REDD': {
				'buildings': [
					'building3': {
						'start-time': '2018-1-1',
						'end-time': '2018-1-2'
					}
				]
			}
		}
	}
}
```

The above mentioned configuration mentions the cases when the users do the training. However, some users might just want to use `pre-trained` model. For such a case, the configuration might look like:

```python
{
	'params': [('power', 'active'), ('power', 'reactive')],
	'sample-rate': 60,
	'appliances': ['fridge', 'air conditioner'],
        'pre-trained': {'FHMM: '/home/abc/fhmm.h5'},
        'methods' : ['FHMM']
	'test': {
		'datasets': {
			'REDD': {
				'buildings': [
					'building3': {
						'start-time': '2018-1-1',
						'end-time': '2018-1-2'
					}
				]
			}
		}
	}
}
```

NILMTK should do the plumbing work behind the scenes. 

Thoughts @PMeira, @OdysseasKr, @MingjunZhong

I am also opening another issue proposal on how we can support algorithm developers.

**Comment by OdysseasKr** [comment]
This sounds very convenient, especially after the discussion at #755 .

Since we want a pipeline that, at least for now, needs to be able to serve:
- the existing disaggregation algorithms
- @MingjunZhong 's models
- Neural NILM models

we should maybe try to gather a list of preprocessing requirements.

**Comment by nipunbatra** [comment]
I am proposing a simple workflow for algorithms requiring preprocessing.

### Specifying the preprocessing via a simple interface

```python
preprocessing =
{
    'params': [('power', 'active')],
    'sample-rate': 60,
    'appliances': ['fridge', 'air conditioner'],
    'chunk-size': 1000,
    'train-store': 'preprocessed-train.h5',
    'test-store': 'preprocessed-test.h5',
    'preprocessing-function': normalise,
    'train': {
        'datasets': {
            'REDD': {
                'buildings': [
                    'building3': {
                        'start-time': '2018-1-1',
                        'end-time': '2018-1-2'
                    }
                ]
            }
        }
    },
    'test': {
        'datasets': {
            'REDD': {
                'buildings': [
                    'building3': {
                        'start-time': '2018-1-1',
                        'end-time': '2018-1-2'
                    }
                ]
            }
        }

    }
}
```

where say we have a normalise function looking like the following:

```python
def normalise(df, *args, **kwargs):
     return new_df
```
and we have a high level API object:

```
api = nilmtk.PreprocessAPI(preprocessing)

# 1. Load data in chunk of preprocessing['chunk-size'] for train
# 2. Call preprocessing function on each of them and store in preprocessing['train-file']

The `train-chunk.h5` is of the format:
keys: /dataset_name/building_name/chunk1
value: preprocessed df

# 3. Repeat # 1 and # 2 for test set

```

and now we can as usual call the high level nilmtk API defined above which will repeatedly make a call to `partial_fit` for training. We can specify a parameter to `randomise` loading of preprocessed chunks. For testing or diaggregating, as usual the nilmtk API takes care of calling the appropriate function. 

The main thing is that the nilmtk API should know how to load and pass around preprocessed data. Currently, it only know how to load and pass around niltmk HDFStore stored dataset. 

In an ideal world, the output of the preprocessing could also be stored into a nilmtk HDFStore format, but, that doesn't seem to be trivial.

**Comment by nipunbatra** [comment]
@Rithwikksvr could you please clarify the issues you were mentioning about requiring a different format for neural methods?

**Comment by Rithwikksvr** [comment]
@nipunbatra Neural methods such as Neural NILM use different sequence lengths for different appliances. But the current method of partial fit supports takes only a single aggregate df (constant sequence length of all appliances) for all disaggregation on all appliances. So, how can we pass multiple aggregate readings, where each of them corresponds to different appliances? 

One solution can be to pass a list of aggregate readings like `(app_name, aggregate_reading)` similar to the list of appliance readings `(app_name, appliance_reading)`. This way we can train on multiple appliances that have different sequence lengths. We just need to pass the corresponding aggregate reading while calling the internal fit for a given appliance. 

Any comments @nipunbatra ?

**Comment by MingjunZhong** [comment]
We should make it clear that in our previous discussions we would like to treat the sequence length as a variable parameter so that users can choose the sequence length. Is it possible to leave it as an input parameter to the API? If so, I would suggest to do so.

**Comment by Rithwikksvr** [comment]
@MingjunZhong we already started making it as an input parameter to the user. But, if the user chooses to use different sequence lengths for each of the appliances, then we also need to store the corresponding aggregate for each for the appliances.

**Comment by OdysseasKr** [comment]
@Rithwikksvr Maybe I am mistaken but I think that the neural methods can only produce results for one appliance at the time. This means that for each appliance passed, a separate neural network needs to be trained. 
- Is training multiple networks simultaneously within the scope?
- Should the disaggregator writer be aware of the requirement for multiple appliance and develop accordingly? Maybe that defeats the purpose of creating a simple partial_fit function.

If the answer for both of the above is yes, then what you propose sounds like a good design.

**Comment by Rithwikksvr** [comment]
@OdysseasKr training of multiple neural networks is within the scope now. We learn a neural network for each of the appliance with the specified parameters. But, learning multiple neural networks where each of the neural nets have a different sequence length won't be possible with the current design. 

Yes the disaggregator writer should be aware. And it surely does defeat the purpose of simple partial_fit.

Instead, we can say that algorithm is currently limited to constant sequence length for all appliances. If the writer wants to try with different sequence length, he/she can do the whole pre-processing and training and testing separately for each appliance by specifying the sequence length. They need to execute the whole pipeline for each of the appliances. I can't see any draw backs with this design.

Any thoughts @nipunbatra

**Comment by nipunbatra** [comment]
@OdysseasKr @MingjunZhong This does sound like a non-trivial problem.

Could you suggest a simple input data structure using just Python data objects: list, Pandas, numpy; that we can use across all methods: both neural methods and the trivial CO, FHMM etc. already developed models.

I think finding such a common data structure or format is the key.

@Rithwikksvr @ayushp20 @raktim2015 @rajatgupta1234 Please feel free to chime in.

**Comment by OdysseasKr** [comment]
My thoughts are that, since the Disaggregator has to train multiple networks, I suggest that the partial fit method receives two parameters:
1. DataFrame containing mains and meter data
2. A dict containing parameters for each appliance. The keys of the dict and the appliance names and each value can be a new dict (or **kwargs) for specific parameters. Eg.
```JSON
{
  "fridge": {
    "window_size": 100
  },
  "washing_machine": {
    "window_size": 140
  }
}
```
The second parameter is optional. If not specified, the Disaggregator uses default values.
The second parameter should be disaggregator specific I guess since it will be different for every model.


The user can specify the parameters in the high-level API using a new attribute as follows:
```Python
preprocessing =
{
    'params': [('power', 'active')],
    'sample-rate': 60,
    'appliances': ['fridge', 'air conditioner'],
    'chunk-size': 1000,
    'train-store': 'preprocessed-train.h5',
    'test-store': 'preprocessed-test.h5',
    'preprocessing-function': normalise,
    'train': {
        'datasets': {
            'REDD': {
                'buildings': [
                    'building3': {
                        'start-time': '2018-1-1',
                        'end-time': '2018-1-2'
                    }
                ]
            }
        }
    },
    'test': {
        'datasets': {
            'REDD': {
                'buildings': [
                    'building3': {
                        'start-time': '2018-1-1',
                        'end-time': '2018-1-2'
                    }
                ]
            }
        }
    },
   'model_params': {
       'fridge': {
           'window_size': 100
        },
        'washing_machine': {
            'window_size': 140
        }
   }
}
```

How does this look?

**Comment by Rithwikksvr** [comment]
@PMeira @nipunbatra where shall we keep the documentation for the API? I have a notebook which demonstrates the use of the API.

One notebook shows the power of API using the NILMTK algorithms such as CO, FHMM, Hart and Mean

The other notebook is for nilmtk-contrib.

**Comment by PMeira** [comment]
@Rithwikksvr I'd say somewhere under `/docs/manual`, either on a dedicated folder or maybe in `/docs/manual/user_guide` with the rest.
